File: golang/src/p22/main.go
package main

import (
  "fmt"
  "io/ioutil"
  "sort"
  "strings"
)

func main() {
  names := readNames()
  sort.Strings(names)

  totalScore := 0
  for i, name := range names {
    wordScore := 0
    for _, c := range name {
      // Only works if names are all capital letters (they are). Add 1 since A is 1, not 0.
      wordScore += int(c - 'A') + 1
    }
    totalScore += wordScore * (i+1)
  }

  fmt.Println(totalScore)
}

func readNames() []string {
  // The file is all names enclosed in quotes with commas on one line:
  // "NAME1","NAME2",...
  data, err := ioutil.ReadFile("../data/p022_names.txt")
  if err != nil {
    panic("Cannot open ../data/p022_names.txt: " + err.Error())
  }

  // First, split on comma
  names := strings.Split(string(data), ",")

  // Then, remove the quotes
  for i, n := range names {
    names[i] = strings.Trim(n, "\"")
  }

  return names
}

File: cpp/src/p22/main.cc
#include <fstream>
#include <iostream>
#include <set>

std::set<std::string> readNames() {
  std::ifstream ifs("../data/p022_names.txt");
  if (!ifs.good()) {
    std::cout << "Cannot read ../data/p022_names.txt\n";
    exit(1);
  }

  // The file is one line with names in all caps surrounded in quotes, split by commas.
  // "NAME1","NAME2",...
  std::set<std::string> names;
  std::string name;
  while (std::getline(ifs, name, ',')) {
    // Remove the quotes on either side of the name (first and last char)
    names.emplace(name.substr(1, name.size()-2));
  }

  return names;
}

int main() {
  unsigned int totalScore = 0;
  const std::set<std::string> names = readNames();
  unsigned int i = 0;
  for (const auto& name : names) {
    int wordScore = 0;
    for (auto c : name) {
      // Only works if names are all capital letters (they are). Add 1 since A is 1, not 0.
      wordScore += c - 'A' + 1;
    }
    ++i;
    totalScore += wordScore * i;
  }

  std::cout << totalScore << "\n";
}
